File: bindings/wasm/poi_wasm/examples/src/04_create_and_verify_object_proof.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * # Create and Verify an Object Proof
 *
 * Starting from only an object ID, the builder fetches the object's latest
 * version, discovers the transaction that produced it, and packages the object
 * and transaction evidence into one proof.
 *
 * The discovered transaction supports the object claim but does not become an
 * explicit transaction target. Trusted-node resolution keeps this focused
 * example fast and is appropriate only when the node is inside the trust boundary.
 */

import { strict as assert } from "node:assert";

import { fromHex, normalizeIotaObjectId } from "@iota/iota-sdk/utils";
import { CommitteeResolution, PoiClient } from "@iota/poi-wasm";
import { STAKED_IOTA_OBJECT_ID } from "./util.js";

/** Demonstrates target-driven object discovery and trusted-node verification. */
export async function createAndVerifyObjectProof(): Promise<void> {
    console.log("=== Proof of Inclusion: Create and Verify an Object Proof ===\n");

    const client = PoiClient.mainnet();

    console.log("Network:       mainnet");
    console.log(`Object target: ${STAKED_IOTA_OBJECT_ID}\n`);

    // No transaction digest is supplied. The builder discovers the transaction
    // that produced the latest object version and constructs its evidence.
    const object = fromHex(normalizeIotaObjectId(STAKED_IOTA_OBJECT_ID, false, true));
    const proof = await client.proof().object(object).build();
    const targets = proof.targets;

    assert.equal(targets.transaction, undefined);
    assert.equal(targets.objects.length, 1);
    assert.equal(targets.objects[0]?.objectId, STAKED_IOTA_OBJECT_ID);
    assert.equal(targets.events.length, 0);

    console.log("Proof constructed:");
    console.log(`  checkpoint epoch: ${proof.checkpointEpoch}`);
    console.log(`  object version:   ${targets.objects[0]?.version}`);
    console.log(`  object targets:   ${targets.objects.length}\n`);

    // This avoids the genesis committee walk but places the selected node inside the trust boundary.
    await client.verifier(CommitteeResolution.trustedNode()).verify(proof);

    console.log("Object proof verified successfully.");
    console.log("The resolved object version was changed by a transaction in the verified checkpoint.");
}

File: examples/poi/04_create_and_verify_object_proof.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! # Create and Verify an Object Proof
//!
//! An application can request a Proof of Inclusion using only an object ID. The
//! builder fetches the object's latest version, discovers the transaction that
//! produced that version, and packages the object and transaction evidence into
//! one proof.
//!
//! The discovered transaction is evidence supporting the object claim. It is not
//! an explicit transaction target unless the caller also invokes `transaction`.
//!
//! This focused example uses trusted-node committee resolution to avoid an epoch
//! walk. Use this mode only when the connected node is inside the verifier's
//! trust boundary.

use anyhow::{Context, Result, ensure};
use poi_examples::prepare_poi_example;
use poi_rs::CommitteeResolution;

/// Demonstrates how to:
/// 1. Request a proof using only an object ID.
/// 2. Let the builder resolve the latest object version and its transaction.
/// 3. Distinguish supporting transaction evidence from an explicit target.
/// 4. Resolve the committee from a trusted node and verify the object claim.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Proof of Inclusion: Create and Verify an Object Proof ===\n");

    let context = prepare_poi_example().await?;
    let object_id = context.create_notarization("PoI object-proof example").await?.object_id;
    let client = &context.poi_client;

    println!("Network:       {}", context.network_alias);
    println!("Object target: {object_id}\n");

    // No transaction digest is supplied. The builder fetches the object first,
    // reads its previous transaction, and constructs evidence for that execution.
    let proof = client
        .proof()
        .object(object_id)
        .build()
        .await
        .context("failed to construct the object proof")?;

    ensure!(
        proof.targets().transaction.is_none(),
        "the discovered transaction must not become an explicit transaction target"
    );
    ensure!(
        proof.targets().objects.len() == 1,
        "the proof must contain one object target"
    );
    ensure!(
        proof.targets().objects[0].as_inner().object_ref().object_id == object_id,
        "the proof must contain the requested object"
    );
    ensure!(
        proof.targets().events.is_empty(),
        "the object-only proof must not contain event targets"
    );

    println!("Proof constructed:");
    println!(
        "  resolved transaction: {}",
        proof.transaction_proof.transaction.digest()
    );
    println!("  object version:       {}", proof.targets().objects[0].version());
    println!("  checkpoint epoch:     {}", proof.checkpoint_summary.epoch());
    println!("  checkpoint number:    {}\n", proof.checkpoint_summary.sequence_number);

    // Trusted-node resolution accepts the committee reported by the connected
    // node. It avoids the genesis walk but makes that node part of the trust boundary.
    let verifier = client.verifier(CommitteeResolution::TrustedNode);
    verifier
        .verify(&proof)
        .await
        .context("object proof verification failed")?;

    println!("Object proof verified successfully.");
    println!("The resolved object version was changed by a transaction included in the verified checkpoint.");

    Ok(())
}
